clippy_lints/src/loops/infinite_loop.rs

```diff
@@ -128,3 +140,65 @@ impl<'hir> Visitor<'hir> for LoopVisitor<'hir, '_> {
         }
     }
 }
+
+/// Similar to [`FnRetTy`], but reveals the actual type of an `OpaqueDef`.
+enum RetTy<'hir> {
+    DefaultReturn(Span),
+    Return(Ty<'hir>),
+}
+
+impl<'hir> RetTy<'hir> {
+    fn from_fn_ret_ty(cx: &LateContext<'hir>, fn_ret_ty: FnRetTy<'hir>) -> Self {
+        /// Reveal and return the related type of an `opaque`, return `None` if the
+        /// given `ty` is not an `OpaqueDef`.
+        fn inner_<'tcx>(cx: &LateContext<'tcx>, ty: &Ty<'tcx>) -> Option<Ty<'tcx>> {
+            /// Visitor to find the type binding.
+            struct BindingVisitor<'tcx> {
+                res: Option<Ty<'tcx>>,
+            }
+            impl<'tcx> Visitor<'tcx> for BindingVisitor<'tcx> {
+                fn visit_assoc_type_binding(&mut self, type_binding: &'tcx hir::TypeBinding<'tcx>) {
+                    if self.res.is_some() {
+                        return;
+                    }
+                    if let hir::TypeBindingKind::Equality {
+                        term: hir::Term::Ty(ty),
+                    } = type_binding.kind
+                    {
+                        self.res = Some(*ty);
+                    }
+                }
+            }
```

Comment:
Visitors can return `ControlFlow` by setting `type ResultTy = ControlFlow<T>`

Reply:
I tried use this for the other visitor [here](https://github.com/rust-lang/rust-clippy/pull/12421/files#diff-5ca8910c6fae00e31b1d2a2c7d7e82f9a9922f8405b6aa815d4ffce8740ea04fR114) as well, but sadly it doesn't work, I'm assuming it's somehow "conflicting" with those `walk_expr` calls?

Reply:
It would work there. `walk_expr` has the same return type as `visit_expr`. If you really want to clean that part up you can get rid the label and loop depth tracking. `Destination::target_id` will always hold the `HirId` of the loop expression or block it's targeting.

Reply:
after trying out your suggestion, I think the depth tracking is somehow needed in this case below right?

```rust
'outer: loop {
    loop {
        if cond {
            break 'outer;
        }
    }
}
```

'outer loop can be identified as finite when visitor sees the `break` with a matching `target_id`. However during the second iteration, the original logic assumes the `break 'outer` was breaking the current loop because the "depth" was 0.

```rust
if self.loop_depth == 0 || has_same_label() {
    self.is_finite = true;
}
```

Without the depth tracking I can't think of any better way to eliminate this FP, so I'm guessing it should be safe to assume that way?

Reply:
A small change to the example causes the current version to not work:
```rust
'outer: loop {
    loop {
        if cond {
            loop {
                break 'outer;
            }
        }
    }
}
```

The correct implementation would track the `HirId` of every loop and targeted block contained in the checked loop.

---

If you want to fix this feel free, otherwise leave the current code.